crypto/evp/evp_lib.c

```diff
@@ -637,14 +645,23 @@ int EVP_CIPHER_get_key_length(const EVP_CIPHER *cipher)
 
 int EVP_CIPHER_CTX_get_key_length(const EVP_CIPHER_CTX *ctx)
 {
-    int ok;
-    size_t v = ctx->key_len;
-    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    /* Check to see if we've cached this value previously */
+    if (ctx->key_length < 0) {
+        int ok, len;
+        size_t v = ctx->key_len;
+        OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
 
-    params[0] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_KEYLEN, &v);
-    ok = evp_do_ciph_ctx_getparams(ctx->cipher, ctx->algctx, params);
+        params[0] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_KEYLEN, &v);
+        ok = evp_do_ciph_ctx_getparams(ctx->cipher, ctx->algctx, params);
 
-    return ok != 0 ? (int)v : EVP_CTRL_RET_UNSUPPORTED;
+        len = ok != 0 ? (int)v : EVP_CTRL_RET_UNSUPPORTED;
```

Comment:
By accident `EVP_CTRL_RET_UNSUPPORTED` is -1 and so storing it in `ctx->key_length` happens to do the right thing. But it right to be relying on this?

Reply:
I'm only relying on EVP_CTRL_RET_UNSUPPORTED being negative which seems like a safer assumption.

Reply:
I've addressed this and cleaned up the conditionals a little.